### PR TITLE
ENH: update `find_repeats`.

### LINKED ISSUE #673: Statistics Review: find_repeats (Trac #146)
_Original ticket http://projects.scipy.org/scipy/ticket/146 on 2006-04-02 by @rkern, assigned to @rkern._

The function `find_repeats` in file source:trunk/Lib/stats/morestats.py needs review.

Please look over the StatisticsReview guidelines and add your comments below.

### PR DESCRIPTION
**Pull request by aeklant**
closes gh-673

### CODE REVIEW

**Comment by jaimefrio** [comment]
This functionality is equivalent to what `np.unique` provides in numpy >= 1.9, with a pure Python implementation and comparable performance:

```
a = np.random.randint(10, size=1000)
%timeit find_repeats(a)
100000 loops, best of 3: 18.9 µs per loop
%timeit np.unique(a, return_counts=True)
10000 loops, best of 3: 21.2 µs per loop

a = np.random.randint(100, size=10000)
%timeit find_repeats(a)
1000 loops, best of 3: 505 µs per loop
%timeit np.unique(a, return_counts=True)
1000 loops, best of 3: 346 µs per loop
```

Is there any reason to keep the Fortran implementation, or would it make sense to reuse numpy's approach? Internally only `find_repeats` seems to use it.

**Comment by aeklant** [comment]
@jaimefrio I agree with your point. The main reason to keep this function (as far as I can tell) is that it is actually quite handy and its usage widespread. Maybe a more experienced user can weigh in on this.

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -248,11 +250,10 @@ def find_repeats(arr):
 
     Returns
     -------
-    find_repeats : tuple
-        Returns a tuple of two 1-D ndarrays.  The first ndarray are the repeats
-        as sorted, unique values that are repeated in `arr`.  The second
-        ndarray are the counts mapped one-to-one of the repeated values
-        in the first ndarray.
+    values : a tuple of the unique values from the input that are repeated
```
Comment (tacaswell) [inline scipy/stats/stats.py]: This makes the docstrings no longer comply with the numpydoc format.

**Comment by rgommers** [comment]
I'd suggest to now note in the docstring that for numpy >= 1.9 ``np.unique`` provides the same functionality. Once 1.9 is the lowest supported numpy version, we can probably just deprecate ``find_repeats`` or otherwise change the inplementation to ``np.unique(arr, return_counts=True)``. Until that time, I wouldn't bother trying to replace the Fortran code (it works, never caused an issue).

File: scipy/stats/stats.py (hunk 1 above)
Comment (aeklant) [inline scipy/stats/stats.py]: Thanks @tacaswell, I was not aware of that.

**Comment by rgommers** [comment]
I didn't check the above comments on ``np.unique`` in detail; I actually just realized now that this is not the same as ``np.unique(arr, return_counts=True)``. ``find_repeats`` returns *only repeated values*:

    In [3]: stats.find_repeats([1, 1, 2])
    Out[3]: (array([ 1.]), array([2]))

    In [4]: stats.find_repeats(np.arange(4))
    Out[4]: (array([], dtype=float64), array([], dtype=int32))

So I suggest to explain this difference clearly, and remove the note on possible future deprecation.

Other remarks:

- Can you move the tests from ``test_morestats.py`` to ``test_stats.py``?
- Ignore my remark in the comment on ``int32``, that is simply due to the difference between 64/32-bit installs, so no need to explain.

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -264,8 +272,13 @@ def find_repeats(arr):
     (array([ 4., 5.]), array([2, 2], dtype=int32))
 
     """
+    RepeatedResults = namedtuple('RepeatedResults', ('values', 'times'))
```
Comment (rgommers) [inline scipy/stats/stats.py]: ``counts`` is maybe clearer than ``times``?

File: scipy/stats/stats.py [hunk 3]
```diff
@@ -241,6 +241,8 @@ def find_repeats(arr):
     """
     Find repeats and repeat counts.
 
+    ndarray input is flattened.
```
Comment (rgommers) [inline scipy/stats/stats.py]: Can you move this to the description of ``values`` in the Parameters list?

**Comment by aeklant** [comment]
Check; please review. I had moved the tests to `test_stats.py` in a previous commit but I'm deleting the tests from `tests_morestats.py` here to avoid confusion.

**Comment by rgommers** [comment]
LGTM, will merge in a day or so unless @jaimefrio has further comments.

**Comment by jaimefrio** [comment]
I'm pretty sure you can get rid of the Fortran code and have the same functionality implemented in a dozen lines of pure Python with a small, if at all, performance penalty. But that is orthogonal to this PR: I'll see about sending a follow-up one once this is merged.

File: scipy/stats/stats.py [hunk 4]
```diff
@@ -248,11 +248,16 @@ def find_repeats(arr):
...
-    find_repeats : tuple
-        Returns a tuple of two 1-D ndarrays.  The first ndarray are the repeats
-        as sorted, unique values that are repeated in `arr`.  The second
-        ndarray are the counts mapped one-to-one of the repeated values
-        in the first ndarray.
+    values : tuple
+        The unique values from the (flattened) input that are repeated.
+
+    counts : tuple
+        Number of times the corresponding 'value' is repeated.
```
Comment (ev-br) [inline scipy/stats/stats.py]: values and counts are arrays, not tuples

```
>>> a = np.random.random(3)
>>> b = np.tile(a, 2)
>>> find_repeats(b)
(array([ 0.40437341,  0.40848151,  0.72190451]), array([2, 2, 2], dtype=int32))
```

**Comment by aeklant** [comment]
Can this PR get some love? I really just want to cross it from the list of pending issues.

**Comment by rgommers** [comment]
Merged, thanks @aeklant

**Pull request merged**